SmartDeviceLink/SDLEncryptionLifecycleManager.m

```diff
@@ -0,0 +1,241 @@
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_hmiStatusDidChange:) name:SDLDidChangeHMIStatusNotification object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_permissionsDidChange:) name:SDLDidChangePermissionsNotification object:nil];
+
+    return self;
+}
+
+- (void)startWithProtocol:(SDLProtocol *)protocol {
+    _protocol = protocol;
+    
+    @synchronized(self.protocol.protocolDelegateTable) {
+        if (![self.protocol.protocolDelegateTable containsObject:self]) {
+            [self.protocol.protocolDelegateTable addObject:self];
+        }
+    }
+    
+    SDLLogD(@"Starting encryption manager");
+}
+
+- (void)stop {
+    _hmiLevel = SDLHMILevelNone;
+    _permissions = [NSMutableDictionary<SDLPermissionRPCName, SDLPermissionItem *> dictionary];
+    _protocol = nil;
+    
+    SDLLogD(@"Stopping encryption manager");
+    [self sdl_stopEncryptionService];
+}
+
+- (void)sendEncryptedRequest:(SDLRPCRequest *)request withResponseHandler:(SDLResponseHandler)handler {
+    if (!self.protocol || !self.isEncryptionReady) {
+        SDLLogW(@"Encryption manager is not yet ready, wait until after proxy is opened");
+        return;
+    }
+    
+    SDLAsynchronousRPCRequestOperation *op = [[SDLAsynchronousRPCRequestOperation alloc] initWithConnectionManager:self.connectionManager request:request withEncryption:YES responseHandler:handler];
+    
+    [self.rpcOperationQueue addOperation:op];
+}
+
+- (BOOL)isEncryptionReady {
+    return [self.encryptionStateMachine isCurrentState:SDLEncryptionLifecycleManagerStateReady];
+}
+
+- (void)sdl_startEncryptionService {
+    SDLLogV(@"Attempting to start Encryption Service");
+    if (!self.protocol) {
+        SDLLogV(@"Encryption manager is not yet started");
+        return;
+    }
+
+    if (!self.hmiLevel || !self.permissions) {
+        SDLLogV(@"Encryption Manager is not ready to encrypt.");
+        return;
+    }
+    
+    if (![self.hmiLevel isEqualToEnum:SDLHMILevelNone]) {
+        [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStarting];
+    } else {
+        SDLLogE(@"Unable to send encryption start service request\n"
+                "permissions: %@\n"
+                "HMI state must be LIMITED, FULL, BACKGROUND: %@\n",
+                self.permissions, self.hmiLevel);
+    }
+}
+
+- (void)sdl_sendEncryptionStartService {
+    SDLLogD(@"Sending secure rpc start service");
+    [self.protocol startSecureServiceWithType:SDLServiceTypeRPC payload:nil tlsInitializationHandler:^(BOOL success, NSError *error) {
+        if (error) {
+            SDLLogE(@"TLS setup error: %@", error);
+            [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+        }
+    }];
+}
+
+- (void)sdl_stopEncryptionService {
```

Comment:
this is called in `SDLEncryptionLifecycleManager` `stop` method

Reply:
I see that now, it wasn't coming up in Xcode. Still, I don't think that this is correct. If the lifecycle manager is stopped, it should properly unregister. This will just end the service, which isn't correct. The service will be ended elsewhere.